File: apps/api/worldpulse/scoring.py
from dataclasses import asdict, dataclass
from math import isfinite, log2

WEIGHTS = {
    "corroboration": 0.45,
    "reporting_momentum": 0.25,
    "freshness": 0.20,
    "geographic_reach": 0.10,
}


def clamp(value: float | int | None, minimum: float = 0, maximum: float = 100) -> float:
    numeric = float(value or 0)
    if not isfinite(numeric):
        return minimum
    return min(maximum, max(minimum, numeric))


def signal_label(score: float) -> str:
    safe = clamp(score)
    if safe >= 75:
        return "Very strong"
    if safe >= 55:
        return "Strong"
    if safe >= 35:
        return "Building"
    return "Early"


@dataclass(slots=True)
class ScoringInput:
    independent_source_count: int = 0
    affected_country_count: int = 0
    age_hours: float = 168
    articles_per_hour: float = 0
    article_count: int = 0
    coverage_window_hours: float = 0


@dataclass(slots=True)
class ScoreResult:
    score: int
    label: str
    components: dict[str, float]
    input: dict[str, float | int]


def calculate_news_signal(data: ScoringInput) -> ScoreResult:
    source_count = max(0, data.independent_source_count)
    corroboration = clamp((log2(source_count + 1) / log2(9)) * 100)
    country_count = max(0, data.affected_country_count)
    geographic_reach = 0 if country_count <= 1 else clamp(((country_count - 1) / 5) * 100)
    normalized = {
        "corroboration": corroboration,
        "reporting_momentum": clamp((data.articles_per_hour / 3) * 100),
        "freshness": clamp(100 - (data.age_hours / 168) * 100),
        "geographic_reach": geographic_reach,
    }
    components = {name: round(value * WEIGHTS[name], 1) for name, value in normalized.items()}
    score = round(clamp(sum(components.values())))
    return ScoreResult(score, signal_label(score), components, asdict(data))

File: lib/scoring.ts
import type {
  Category,
  SignalLabel,
  ScoringComponents,
  ScoringInput,
} from "./types";

const WEIGHTS = {
  corroboration: 0.45,
  reportingMomentum: 0.25,
  freshness: 0.2,
  geographicReach: 0.1,
} as const;

const CATEGORY_COLORS: Record<Category, string> = {
  Politics: "#a78bfa",
  Economy: "#3b82f6",
  "Conflict and security": "#fb4d5f",
  "Crime and justice": "#f97316",
  Environment: "#22c55e",
  "Weather and disasters": "#f59e0b",
  Health: "#ec4899",
  "Science and technology": "#06b6d4",
  "Society and education": "#818cf8",
  "Travel and transport": "#14b8a6",
  "Culture and entertainment": "#d946ef",
  Sports: "#84cc16",
  "Food and agriculture": "#facc15",
  "Local affairs": "#38bdf8",
  Other: "#94a3b8",
};

const clamp = (value: number, min = 0, max = 100) =>
  Math.min(max, Math.max(min, Number.isFinite(value) ? value : 0));

export function signalLabel(score: number): SignalLabel {
  const safeScore = clamp(score);
  if (safeScore >= 75) return "Very strong";
  if (safeScore >= 55) return "Strong";
  if (safeScore >= 35) return "Building";
  return "Early";
}

export function calculateNewsSignal(input: Partial<ScoringInput>): {
  score: number;
  label: SignalLabel;
  components: ScoringComponents;
} {
  const sourceCount = Math.max(0, input.independentSourceCount ?? 0);
  const corroboration = clamp(
    (Math.log2(sourceCount + 1) / Math.log2(9)) * 100,
  );
  const affectedCountryCount = Math.max(0, input.affectedCountryCount ?? 0);
  const geographicReach =
    affectedCountryCount <= 1
      ? 0
      : clamp(((affectedCountryCount - 1) / 5) * 100);
  const freshness = clamp(100 - ((input.ageHours ?? 168) / 168) * 100);
  const reportingMomentum = clamp(
    ((input.articlesPerHour ?? 0) / 3) * 100,
  );

  const components: ScoringComponents = {
    corroboration:
      Math.round(corroboration * WEIGHTS.corroboration * 10) / 10,
    reportingMomentum:
      Math.round(reportingMomentum * WEIGHTS.reportingMomentum * 10) / 10,
    freshness: Math.round(freshness * WEIGHTS.freshness * 10) / 10,
    geographicReach:
      Math.round(geographicReach * WEIGHTS.geographicReach * 10) / 10,
  };

  const score = Math.round(
    clamp(
      components.corroboration +
        components.reportingMomentum +
        components.freshness +
        components.geographicReach,
    ),
  );

  return { score, label: signalLabel(score), components };
}

function mixHex(foreground: string, background: string, amount: number) {
  const clean = (value: string) => value.replace("#", "");
  const fg = clean(foreground);
  const bg = clean(background);
  const blend = [0, 2, 4]
    .map((index) => {
      const a = Number.parseInt(fg.slice(index, index + 2), 16);
      const b = Number.parseInt(bg.slice(index, index + 2), 16);
      return Math.round(a * amount + b * (1 - amount))
        .toString(16)
        .padStart(2, "0");
    })
    .join("");
  return `#${blend}`;
}

export function mapStyleForEvent(
  category?: Category,
  importanceScore = 0,
): { fillColor: string; fillOpacity: number } {
  if (!category) return { fillColor: "#303a47", fillOpacity: 0.82 };
  const intensity = 0.72 + clamp(importanceScore) * 0.0028;
  return {
    fillColor: mixHex(CATEGORY_COLORS[category], "#182333", intensity),
    fillOpacity: 0.94,
  };
}

export function categoryColor(category: Category) {
  return CATEGORY_COLORS[category];
}

export const scoringWeights = WEIGHTS;
